src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -22,33 +22,35 @@
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.Condition;
-import java.util.concurrent.locks.LockSupport;
 import java.util.concurrent.locks.ReentrantLock;
 
 
 public class TarantoolClientImpl extends TarantoolBase<Future<?>> implements TarantoolClient {
+
     public static final CommunicationException NOT_INIT_EXCEPTION = new CommunicationException("Not connected, initializing connection");
+
     protected TarantoolClientConfig config;
 
     /**
      * External
      */
     protected SocketChannelProvider socketProvider;
+    protected SocketChannel channel;
+    protected ReadableViaSelectorChannel readChannel;
+
     protected volatile Exception thumbstone;
 
     protected Map<Long, TarantoolOp<?>> futures;
-    protected AtomicInteger wait = new AtomicInteger();
+    protected AtomicInteger pendingResponsesCount = new AtomicInteger();
```

Comment:
We use `ConcurrentHashMap` to store the futures. This implementation has some drawback when we call aggegative functions [(ConcurentHashMap)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) 

> Bear in mind that the results of aggregate status methods including size, isEmpty, and containsValue are typically useful only when a map is not undergoing concurrent updates in other threads. Otherwise the results of these methods reflect transient states that may be adequate for monitoring or estimation purposes, but not for program control.

Reply:
Understood now. Okay.